In [2]:
import json

In [3]:
data_path = './benchmark_wiki_counterfact_test_cf.json'
with open(data_path, 'r') as f:
    data_list = json.load(f)

len(data_list)

885

In [ ]:
data_path = './benchmark_ZsRE_ZsRE-test-all.json'
with open(data_path, 'r') as f:
    data_list = json.load(f)

In [1]:
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
import torch
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0'
import json

model_path = '/share/huggingface/Meta-Llama-3-8B-Instruct'
# tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForCausalLM.from_pretrained(model_path, torch_dtype=torch.bfloat16)
model

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(128256, 4096)
    (layers): ModuleList(
      (0-31): 32 x LlamaDecoderLayer(
        (self_attn): LlamaSdpaAttention(
          (q_proj): Linear(in_features=4096, out_features=4096, bias=False)
          (k_proj): Linear(in_features=4096, out_features=1024, bias=False)
          (v_proj): Linear(in_features=4096, out_features=1024, bias=False)
          (o_proj): Linear(in_features=4096, out_features=4096, bias=False)
          (rotary_emb): LlamaRotaryEmbedding()
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=4096, out_features=14336, bias=False)
          (up_proj): Linear(in_features=4096, out_features=14336, bias=False)
          (down_proj): Linear(in_features=14336, out_features=4096, bias=False)
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm()
        (post_attention_layernorm): LlamaRMSNorm()
      )
    )
    (norm): LlamaRMSNorm()
  )
  (lm_head)

model.embed_tokens.weight torch.Size([128256, 4096])
model.layers.0.self_attn.q_proj.weight torch.Size([4096, 4096])
model.layers.0.self_attn.k_proj.weight torch.Size([1024, 4096])
model.layers.0.self_attn.v_proj.weight torch.Size([1024, 4096])
model.layers.0.self_attn.o_proj.weight torch.Size([4096, 4096])
model.layers.0.mlp.gate_proj.weight torch.Size([14336, 4096])
model.layers.0.mlp.up_proj.weight torch.Size([14336, 4096])
model.layers.0.mlp.down_proj.weight torch.Size([4096, 14336])
model.layers.0.input_layernorm.weight torch.Size([4096])
model.layers.0.post_attention_layernorm.weight torch.Size([4096])

model.norm.weight torch.Size([4096])
lm_head.weight torch.Size([128256, 4096])

In [3]:
for k,v in model.named_parameters():
    print(k, v.shape)

model.embed_tokens.weight torch.Size([128256, 4096])
model.layers.0.self_attn.q_proj.weight torch.Size([4096, 4096])
model.layers.0.self_attn.k_proj.weight torch.Size([1024, 4096])
model.layers.0.self_attn.v_proj.weight torch.Size([1024, 4096])
model.layers.0.self_attn.o_proj.weight torch.Size([4096, 4096])
model.layers.0.mlp.gate_proj.weight torch.Size([14336, 4096])
model.layers.0.mlp.up_proj.weight torch.Size([14336, 4096])
model.layers.0.mlp.down_proj.weight torch.Size([4096, 14336])
model.layers.0.input_layernorm.weight torch.Size([4096])
model.layers.0.post_attention_layernorm.weight torch.Size([4096])
model.layers.1.self_attn.q_proj.weight torch.Size([4096, 4096])
model.layers.1.self_attn.k_proj.weight torch.Size([1024, 4096])
model.layers.1.self_attn.v_proj.weight torch.Size([1024, 4096])
model.layers.1.self_attn.o_proj.weight torch.Size([4096, 4096])
model.layers.1.mlp.gate_proj.weight torch.Size([14336, 4096])
model.layers.1.mlp.up_proj.weight torch.Size([14336, 4096])
model.

glm生成辅助数据

In [1]:
# TODO:暂未考虑主体同义替换
n = 10
system_prompt = f"""You are an expert in sentence rewriting. Your task is to perform a synonymous rewriting of the input sentence and output {n} rewritten sentences.

Requirements:
1. The semantic meaning of the rewritten sentences must strictly match the original sentence.
2. Use different grammatical structures and expressions to rephrase the input sentence as much as possible.
3. Output format requirements: separate the {n} rewritten sentences with a newline (\n)."""

In [2]:
import json
from zhipuai import ZhipuAI
client = ZhipuAI(api_key="1239bfa4ba1cc7135d4476e3ad30af96.JjzOdCR7qNwFSZk7")

In [ ]:
data_path = './benchmark_ZsRE_ZsRE-test-all.json'
with open(data_path, 'r') as f:
    data_list = json.load(f)

with open('ZsRE-test-all-generate-10.json', 'a') as f:
    for data in data_list[1185:]:
        user_prompt = data['prompt']
        response = client.chat.completions.create(
            model="glm-4",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
        )
        data['generate_prompt'] = response.choices[0].message.content.split('\n')
        f.write(json.dumps(data, ensure_ascii=False) + '\n')
        f.flush()

In [3]:
data_path = './benchmark_wiki_counterfact_test_cf.json'
with open(data_path, 'r') as f:
    data_list = json.load(f)

with open('wiki_counterfact_test_cf-generate-10.json', 'a') as f:
    for data in data_list[601:]:
        user_prompt = data['prompt']
        response = client.chat.completions.create(
            model="glm-4",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
        )
        data['generate_prompt'] = response.choices[0].message.content.split('\n')
        f.write(json.dumps(data, ensure_ascii=False) + '\n')
        f.flush()

In [ ]:
user_prompt

测试llm输出稳定

In [ ]:
import transformers
import torch
import os
os.environ['CUDA_VISIBLE_DEVICES']='0'

In [ ]:
# model_id = "meta-llama/Meta-Llama-3-8B"
model_id = "/share/huggingface/Meta-Llama-3-8B-Instruct"

pipeline = transformers.pipeline("text-generation", 
                                 model=model_id, 
                                 model_kwargs={"torch_dtype": torch.bfloat16}, 
                                 device_map="auto")
pipeline("The name of the country of citizenship of Leonardo DiCaprio is")

In [ ]:
pipeline('Which family does Epaspidoceras belong to?')

In [ ]:
model_id = "/share/huggingface/gpt-j-6b"

pipeline = transformers.pipeline("text-generation", 
                                 model=model_id, 
                                 model_kwargs={"torch_dtype": torch.bfloat16}, 
                                 device_map="auto")
pipeline("The name of the country of citizenship of Leonardo DiCaprio is")

In [ ]:
pipeline('Which family does Epaspidoceras belong to?')

In [ ]:
! pip install ipywidgets

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
import torch
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '2'
import json

model_path = '/share/huggingface/gpt-j-6b'
tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForCausalLM.from_pretrained(model_path, torch_dtype=torch.bfloat16)
prompt = "The name of the country of citizenship of Leonardo DiCaprio is"

pipleline_model = pipeline('text2text-generation', model=model, tokenizer=tokenizer)

pipleline_model(prompt, max_length=100, num_return_sequences=1)